service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/utils/DateFormatUtil.java

```diff
@@ -1,26 +1,73 @@
+import static ca.uhn.fhir.model.api.TemporalPrecisionEnum.MINUTE;
+import static ca.uhn.fhir.model.api.TemporalPrecisionEnum.MONTH;
+import static ca.uhn.fhir.model.api.TemporalPrecisionEnum.YEAR;
+
+import ca.uhn.fhir.model.api.TemporalPrecisionEnum;
+
+import java.time.LocalDateTime;
 import java.time.ZoneId;
 import java.time.format.DateTimeFormatter;
-import java.time.format.DateTimeFormatterBuilder;
-import java.util.Date;
+import java.util.Optional;
+
+import org.hl7.fhir.dstu3.model.DateTimeType;
+import org.hl7.fhir.dstu3.model.DateType;
+import org.hl7.fhir.dstu3.model.InstantType;
+
+import com.google.common.collect.ImmutableMap;
 
 import uk.nhs.adaptors.gp2gp.ehr.exception.EhrMapperException;
 
 public class DateFormatUtil {
-    private static final String UK_ZONE_ID = "Europe/London";
-    private static final DateTimeFormatter DATE_TIME_FORMATTER = new DateTimeFormatterBuilder()
-        .appendPattern("yyyyMMddHHmmss")
-        .toFormatter();
-
-    public static String formatDate(Date date) {
-        if (date == null) {
-            throw new EhrMapperException("Could not format date");
+
+    public static final String UK_TIMEZONE = "Europe/London";
+    public static final String COULD_NOT_FORMAT_DATE = "Could not format date";
+    private static final ImmutableMap<TemporalPrecisionEnum, String> YEAR_PATTERN_MAP = ImmutableMap.of(
+        YEAR, "yyyy",
+        MONTH, "yyyyMM",
+        DAY, "yyyyMMdd",
+        MINUTE, "yyyyMMddHHmm"
+    );
+
+    public static String formatDateType(DateType dateType) {
+        if (!dateType.hasValue()) {
+            throw new EhrMapperException(COULD_NOT_FORMAT_DATE);
+        }
+
+        return Optional.ofNullable(YEAR_PATTERN_MAP.get(dateType.getPrecision()))
+            .map(pattern -> dateTypeConverter(dateType, pattern))
+            .orElseGet(() -> dateTypeConverter(dateType, "yyyyMMddHHmmss"));
+    }
+
+    public static String formatInstantType(InstantType dateInstantType) {
+        if (!dateInstantType.hasValue()) {
+            throw new EhrMapperException(COULD_NOT_FORMAT_DATE);
         }
 
-        return DATE_TIME_FORMATTER.format(
-            date.toInstant()
-                .atZone(ZoneId.of(UK_ZONE_ID))
-                .toLocalDateTime());
+        DateTimeType dateTimeType = new DateTimeType(dateInstantType.toCalendar().toInstant().toString());
+
+        return formatDateTimeType(dateTimeType);
+    }
+
+    public static String formatDateTimeType(DateTimeType dateTimeType) {
+        if (!dateTimeType.hasValue()) {
+            throw new EhrMapperException(COULD_NOT_FORMAT_DATE);
+        }
+
+        return Optional.ofNullable(YEAR_PATTERN_MAP.get(dateTimeType.getPrecision()))
+            .map(pattern -> dateTimeTypeConverter(dateTimeType, pattern))
+            .orElseGet(() -> dateTimeTypeConverter(dateTimeType, "yyyyMMddHHmmss"));
+    }
+
+    private static String dateTypeConverter(DateType effectiveDateTimeType, String pattern) {
```

Comment:
Consider replacing these two methods with this:

```
    private static String getFormatStringForPrecision(TemporalPrecisionEnum precisionEnum) {
        return YEAR_PATTERN_MAP.getOrDefault(precisionEnum, "yyyyMMddHHmmss");
    }

    private static String convertWithPrecision(TemporalPrecisionEnum precisionEnum, Calendar calendar) {
        return LocalDateTime.ofInstant(calendar.toInstant(), ZoneId.of(UK_TIMEZONE))
            .format(DateTimeFormatter.ofPattern(getFormatStringForPrecision(precisionEnum)));
    }
```

Then each FHIR-type specific method is only concerned with getting the Calendar object from HAPI's *Type. The implementation would be slightly different if you've already created the formatters as suggested above.